chainbase/src/main/java/org/tron/core/service/RewardCalService.java

```diff
@@ -0,0 +1,258 @@
+
+
+  @Autowired
+  public RewardCalService(@Autowired  DynamicPropertiesStore propertiesStore,
+      @Autowired DelegationStore delegationStore, @Autowired AccountStore accountStore) {
+    this.propertiesStore = propertiesStore.getDb();
+    this.delegationStore = delegationStore.getDb();
+    this.accountStore = accountStore.getDb();
+  }
+
+  @PostConstruct
+  private void init() throws IOException {
+    this.newRewardCalStartCycle = this.getNewRewardAlgorithmEffectiveCycle();
+    if (newRewardCalStartCycle != Long.MAX_VALUE) {
+      isDoneKey = ByteArray.fromLong(newRewardCalStartCycle);
+      if (rewardCacheStore.has(isDoneKey)) {
+        logger.info("RewardCalService is already done");
+        return;
+      }
+      calReward();
+    }
+  }
+
+  @PreDestroy
+  private void destroy() {
+    es.shutdownNow();
+  }
+
+  public void calReward() throws IOException {
+    initLastAccount();
+    es.submit(this::startRewardCal);
+  }
+
+  public void calRewardForTest() throws IOException {
+    this.newRewardCalStartCycle = this.getNewRewardAlgorithmEffectiveCycle();
+    isDoneKey = ByteArray.fromLong(newRewardCalStartCycle);
+    if (rewardCacheStore.has(isDoneKey)) {
+      logger.info("RewardCalService is already done");
+      return;
+    }
+    initLastAccount();
+    startRewardCal();
+  }
+
+  private void initLastAccount() throws IOException {
+    try (DBIterator iterator = rewardCacheStore.iterator()) {
+      iterator.seekToLast();
+      if (iterator.valid()) {
+        byte[] key  = iterator.getKey();
+        System.arraycopy(key, 0, lastAccount, 0, ADDRESS_SIZE);
+      }
+    }
+  }
+
+
+  private void startRewardCal() {
+    if (!doing.compareAndSet(false, true)) {
+      logger.info("RewardCalService is doing");
+      return;
+    }
+    logger.info("RewardCalService start from lastAccount: {}", ByteArray.toHexString(lastAccount));
+    DBIterator iterator = (DBIterator) accountStore.iterator();
+    iterator.seek(lastAccount);
+    iterator.forEachRemaining(e -> {
+      try {
+        doRewardCal(e.getKey(), e.getValue());
+      } catch (InterruptedException error) {
+        Thread.currentThread().interrupt();
+      }
+    });
+    rewardCacheStore.put(this.isDoneKey, IS_DONE_VALUE);
+    logger.info("RewardCalService is done");
+  }
+
+  private void doRewardCal(byte[] address, byte[] account) throws InterruptedException {
+    List<Protocol.Vote> votesList = this.parseVotesList(account);
+    if (votesList.isEmpty()) {
+      return;
+    }
+    long beginCycle = this.getBeginCycle(address);
+    if (beginCycle >= newRewardCalStartCycle) {
+      return;
+    }
+    long endCycle = this.getEndCycle(address);
+    if (endCycle >= newRewardCalStartCycle) {
+      return;
+    }
+    //skip the last cycle reward
+    boolean skipLastCycle = beginCycle + 1 == endCycle;
+    if (skipLastCycle) {
+      beginCycle += 1;
+    }
+    if (beginCycle >= newRewardCalStartCycle) {
+      return;
```

Comment:
In this scenario `(beginCycle + 1 == endCycle)`, is it possible to avoid calculating?